### PR TITLE
Feat/migrate global config to pydantic

### PR DESCRIPTION
**Pull request by petioptrv**
**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

This PR refactors the global configuration map to the new Pydantic approach of configuration management. It also adapts all parts of the code that use the global config. 

In addition, the config is no longer accessed as a singleton, but it is instead passed to the parts of the code that require it. To reflect this change, the config map was renamed to `client_config_map`. Also, to achieve this, the connectors had to be refactored to now accept the client config map as a parameter and store it as an attribute. To indicate to the connectors and any other parts of the code that have to access a config, but must be ensure to not modify it, the `ReadOnlyClientConfigAdapter` was added.

**Tests performed by the developer**

- Adds unit tests for the global config refactor
- Manually tested
    - Creating a PMM config (still legacy) and running the strategy
    - Creating an AMM config and running the strategy
    - Creating a XEMM config and running the strategy
    - Changing global config
    - Visual inspection of the config files for all of the above mentioned configs

**Tips for QA testing**:

Please test the following:
- That all connectors are able to initialize properly (anything past initialization is not affected)
- That legacy strategy configs can be created and started properly (e.g. PMM)
- That AMM, and XEMM can be configured and started properly
- That the global config can be configured and that the various settings are respected by the client

### CODE REVIEW

**Comment by rxlxrxsx** [comment]
For clarification - Normally when entering an incorrect API key the fields **Keys Added** shows YES and **Keys Confirmed** shows NO however under this branch I observed that both fields just shows NO and the .yml is not saved as well. Is this expected?

Screen Shot 2022-06-03 at 1 20 38 PM

**Comment by rxlxrxsx** [comment]
When configuring the color primary_label the prompt shows `secondary` instead of `primary`

Screen Shot 2022-06-03 at 2 54 23 PM

**Comment by petioptrv** [comment]
@gerald-hb, @rxlxrxsx, I have addressed all of the flagged issues in the latest push. Re the behaviour on failure to verify the credentials for an exchange, I have restored the previous functionality.

**Comment by Deepali6997** [comment]
hi @petioptrv all the addressed issues have been resolved.

Screenshot 2022-06-03 at 5 23 18 PM
Screenshot 2022-06-03 at 5 23 31 PM
Screenshot 2022-06-03 at 5 23 42 PM

**Comment by rxlxrxsx** [comment]
Getting this error after entering the `gateway create` command

Screen Shot 2022-06-03 at 10 17 55 PM

```python
2022-06-03 22:17:42,896 - 43615 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: _http_client() missing 1 required positional argument: 'client_config_map'
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/Users/reiniersalazar/hummingbot/hummingbot/client/command/gateway_command.py", line 173, in _create_gateway
    await self._generate_certs(from_client_password=True)  # create cert
  File "/Users/reiniersalazar/hummingbot/hummingbot/client/command/gateway_command.py", line 117, in _generate_certs
    self._get_gateway_instance().reload_certs()
  File "/Users/reiniersalazar/hummingbot/hummingbot/core/gateway/gateway_http_client.py", line 92, in reload_certs
    cls._http_client(re_init=True)
TypeError: _http_client() missing 1 required positional argument: 'client_config_map'
```

**Comment by RC-13** [comment]
I saw the same issue mentioned by Rein in PR#252.

**Comment by dennisocana** [comment]
[sc-21827]

**Comment by petioptrv** [comment]
@rxlxrxsx, @RC-13, my latest commit fixes the bug you've flagged.

**Comment by gerald-hb** [comment]
Gateway create now works as expected but we are hitting another issue on the balance command.
image
Test were done on AWS cloud with the latest commit pushed. Tested with 2 wallet address but getting the same result.

**Comment by RC-13** [comment]
@gerald-hb This issue is not introduced in this PR, I saw this issue reported before on some gateway PRs.

File: hummingbot/client/config/client_config_map.py [hunk 1]
```diff
@@ -0,0 +1,851 @@
...
+        return kill_switch
+
+
+KILL_SWITCH_MODES = {
+    KillSwitchEnabledMode.Config.title: KillSwitchEnabledMode,
+    KillSwitchDisabledMode.Config.title: KillSwitchDisabledMode,
+}
+
+
+class AutofillImportEnum(str, ClientConfigEnum):
```
Comment (aarmoa) [inline hummingbot/client/config/client_config_map.py:229]: Our of curiosity: why do we add the suffix "Enum" to Enum classes?

File: hummingbot/client/config/client_config_map.py [hunk 2]
```diff
@@ -0,0 +1,851 @@
...
+        ),
+    )
+    gateway: GatewayConfigMap = Field(
+        default=GatewayConfigMap(),
+        description=("Gateway API Configurations"
+                     "\ndefault host to only use localhost"
+                     "\nPort need to match the final installation port for Gateway"),
+    )
+    anonymized_metrics_mode: Union[tuple(METRICS_MODES.values())] = Field(
+        default=AnonymizedMetricsDisabledMode(),
```
Comment (aarmoa) [inline hummingbot/client/config/client_config_map.py:676]: This default value is incorrect. The current default value for this field is metrics enabled.

File: hummingbot/client/ui/hummingbot_cli.py [hunk 3]
```diff
@@ -98,15 +101,22 @@ def __init__(self,
...
+        self.app = Application(
+            layout=self.layout,
+            full_screen=True,
+            key_bindings=self.bindings,
+            style=load_style(self.client_config_map),
+            mouse_support=True,
+            clipboard=PyperclipClipboard(),
+        )
         await self.app.run_async(pre_run=self.did_start_ui)
+        # await self.app.run_async(pre_run=partial(self.did_start_ui, self.app))
```
Comment (aarmoa) [inline hummingbot/client/ui/hummingbot_cli.py:119]: Is this required or should it be removed?

File: hummingbot/connector/connector_base.pyx [hunk 4]
```diff
@@ -17,6 +15,14 @@ from hummingbot.core.network_iterator import NetworkIterator
 from hummingbot.core.rate_oracle.rate_oracle import RateOracle
 from hummingbot.core.utils.estimate_fee import estimate_fee
 
+if TYPE_CHECKING:
+    from hummingbot.client.config.client_config_map import ClientConfigMap
+    from hummingbot.client.config.config_helpers import ClientConfigAdapter
+
+NaN = float("nan")
+s_decimal_NaN = Decimal("nan")
+s_decimal_0 = Decimal(0)
```
Comment (aarmoa) [inline hummingbot/connector/connector_base.pyx:24]: Why do we need this if we also have the following import line (line 8)?
`from hummingbot.connector.constants import NaN, s_decimal_NaN, s_decimal_0`

File: hummingbot/connector/exchange/paper_trade/paper_trade_exchange.pyx [hunk 5]
```diff
@@ -1036,6 +1045,9 @@ cdef class PaperTradeExchange(ExchangeBase):
         else:
             return Decimal(f"1e-10")
 
+    def get_order_price_quantum(self, trading_pair: str, price: Decimal) -> Decimal:
+        return self.c_get_order_price_quantum(trading_pair, price)
+
```
Comment (aarmoa) [inline hummingbot/connector/exchange/paper_trade/paper_trade_exchange.pyx:1050]: Is this part of the config map migration?

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 6]
```diff
@@ -867,14 +867,7 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
...
 
-            try:
-                taker_price = taker_market.c_get_vwap_for_volume(taker_trading_pair, False, size).result_price
-            except ZeroDivisionError:
-                return s_decimal_nan
-
-            # If quote assets are not same, convert them from taker's quote asset to maker's quote asset
-            if market_pair.maker.quote_asset != market_pair.taker.quote_asset:
-                taker_price *= self.market_conversion_rate()
+            taker_price = self.c_calculate_effective_hedging_price(market_pair, is_bid, size)
```
Comment (aarmoa) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:870]: Are the changes in XEMM module related to config migration?

**Review by aarmoa — CHANGES_REQUESTED** [review]
I have a couple of comments and some questions.

File: hummingbot/client/config/client_config_map.py (hunk 2 above)
Comment (petioptrv) [reply hummingbot/client/config/client_config_map.py:676]: Addressed

File: hummingbot/client/ui/hummingbot_cli.py (hunk 3 above)
Comment (petioptrv) [reply hummingbot/client/ui/hummingbot_cli.py:119]: Removed

File: hummingbot/connector/connector_base.pyx (hunk 4 above)
Comment (petioptrv) [reply hummingbot/connector/connector_base.pyx:24]: Removed

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx (hunk 6 above)
Comment (petioptrv) [reply hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:870]: No, this is something that I noticed while investigating an issue with XEMM. The code I've removed is a duplicate of `c_calculate_effective_hedging_price`.

File: hummingbot/connector/exchange/paper_trade/paper_trade_exchange.pyx (hunk 5 above)
Comment (petioptrv) [reply hummingbot/connector/exchange/paper_trade/paper_trade_exchange.pyx:1050]: I think I've originally added that for debugging, but without it, `get_order_price_quantum` is undefined

File: hummingbot/client/config/client_config_map.py (hunk 1 above)
Comment (aarmoa) [reply hummingbot/client/config/client_config_map.py:229]: No problem, let's keep the name until we come up with a better alternative.

**Pull request merged**